File: apps/tablet/lib/services/devices/scale_adapter.dart
class WeightReading {
  const WeightReading({
    required this.grossWeight,
    required this.unit,
    required this.isStable,
    required this.rawPayload,
  });

  final String grossWeight;
  final String unit;
  final bool isStable;
  final String rawPayload;
}

abstract interface class ScaleAdapter {
  Future<List<String>> discover();
  Future<void> connect(String deviceIdentifier);
  Future<void> disconnect();
  Stream<WeightReading> readings();
  Future<void> zero();
  Future<void> tare();
}

File: apps/platform/app/Domain/Devices/Contracts/ScaleAdapter.php
<?php

namespace App\Domain\Devices\Contracts;

use App\Domain\Devices\Data\DeviceConnection;
use App\Domain\Devices\Data\WeightReading;

interface ScaleAdapter
{
    /** @return array<int, DeviceConnection> */
    public function discover(): array;

    public function connect(string $deviceIdentifier): DeviceConnection;

    public function disconnect(string $deviceIdentifier): void;

    public function read(string $deviceIdentifier): WeightReading;

    public function zero(string $deviceIdentifier): void;

    public function tare(string $deviceIdentifier): void;
}
